### PR TITLE
BUG: Prevent hang traversing ufunc userloop linked list

### PR DESCRIPTION
**Pull request by eric-wieser**
Fixes gh-9351

### CODE REVIEW

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 1]
```diff
@@ -1792,8 +1792,12 @@ type_tuple_userloop_type_resolver(PyUFuncObject *self,
...
                 continue;
             }
-            funcdata = (PyUFunc_Loop1d *)NpyCapsule_AsVoidPtr(obj);
-            while (funcdata != NULL) {
+
+            for (
+                funcdata = (PyUFunc_Loop1d *)NpyCapsule_AsVoidPtr(obj);
+                funcdata != NULL;
+                funcdata = funcdata->next
+            ) {
```
Comment (charris) [inline numpy/core/src/umath/ufunc_type_resolution.c]: The `)` should probably go on the preceding line. Maybe something like
```
            for (funcdata = (PyUFunc_Loop1d *)NpyCapsule_AsVoidPtr(obj);
                 funcdata != NULL;
                 funcdata = funcdata->next) {
```
If the initialization is put on the previous line, the rest will fit on one.

**Comment by charris** [comment]
Style nit, otherwise looks good to me.

**Comment by charris** [comment]
I went ahead and put a backport-candidate label on it, might as well fix some bugs while waiting on Python 3.6.2.

**Comment by charris** [comment]
I can't think of any sensible way to test this. Not working is one thing, hanging forever is another ;)

**Comment by charris** [comment]
Oops, didn't mean to close this.

File: numpy/core/src/umath/ufunc_type_resolution.c (hunk 1 above)
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_type_resolution.c]: I'd much prefer the initialization to remain in the `for`, but I'm happy to change the wrapping to match the above

**Comment by eric-wieser** [comment]
Found a test using `rational` that still fails after this patch - this wasn't the only place a linked list was not iterated safely

**Comment by eric-wieser** [comment]
Test added, other bug fixed, nit fixed, and rebased on the 1.13 branch point for easy backporting

File: numpy/core/src/umath/ufunc_type_resolution.c [hunk 2]
```diff
@@ -1305,8 +1305,9 @@ find_userloop(PyUFuncObject *ufunc,
             if (obj == NULL) {
                 continue;
             }
-            funcdata = (PyUFunc_Loop1d *)NpyCapsule_AsVoidPtr(obj);
-            while (funcdata != NULL) {
+            for (funcdata = (PyUFunc_Loop1d *)NpyCapsule_AsVoidPtr(obj);
```
Comment (charris) [inline numpy/core/src/umath/ufunc_type_resolution.c]: I'm thinking we actually want to indent these more, like we do with extended if conditions. Should probably put it in the style guide. Anyway, I'm not going to worry about that now.

**Comment by charris** [comment]
Thanks Eric.

**Pull request merged**